pkg/controller/attach_detach.go

```diff
@@ -587,154 +572,160 @@ func reportError(azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, state azd
+	return nil
+}
 
-				for k, v := range desiredAzVolumeAttachment.Annotations {
-					azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+func (r *ReconcileAttachDetach) loadOrStoreVolumeInfo(ctx context.Context, volumeName, azVolumeAttachmentName string) (string, map[string]string, *volumeInfo, bool, error) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	var err error
+	var volumeHandle string
+	var volumeContext map[string]string
+	hasSynced := false
+	vInfo, ok := r.volumeInfos[volumeName]
+	if ok {
+		if synced, hasAzVolumeAttachment := vInfo.syncedAzVolumeAttachments[azVolumeAttachmentName]; hasAzVolumeAttachment && synced {
+			hasSynced = true
+			return volumeHandle, volumeContext, vInfo, hasSynced, nil
+		}
+		volumeHandle = vInfo.handle
+		volumeContext = vInfo.attributes
+	} else {
+		w.Logger().Infof("getting volumeHandle and volumeContext from PersistentVolume for volume (%s)", volumeName)
+		volumeHandle, volumeContext, err = r.getVolumeInfoFromPV(ctx, volumeName)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to find volume info from PersistentVolume for volume (%s)", volumeName)
+			return "", nil, vInfo, hasSynced, err
+		}
+		vInfo = newVolumeInfo(volumeHandle, volumeContext)
+		r.volumeInfos[volumeName] = vInfo
+	}
+	return volumeHandle, volumeContext, vInfo, hasSynced, nil
+}
 
-				azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Update(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			} else {
-				statusUpdateRequired = false
-			}
+func newVolumeInfo(volumeHandle string, volumeContext map[string]string) *volumeInfo {
+	var vInfo volumeInfo
+	vInfo.handle = volumeHandle
+	vInfo.attributes = volumeContext
+	vInfo.syncedAzVolumeAttachments = make(map[string]bool)
+	return &vInfo
+}
 
-			if statusUpdateRequired {
-				azVolumeAttachment.Status.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, consts.VolumeAttachmentKey, volumeAttachment.Name)
+func (r *ReconcileAttachDetach) recreateAzVolumeAttachments(ctx context.Context) error {
+	var err error
+	w, _ := workflow.GetWorkflowFromContext(ctx)
 
-				// update status
-				_, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).UpdateStatus(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update status of AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			}
+	// try to recreate missing AzVolumeAttachment CRI with Primary role from VolumeAttachment
+	for i := 0; i < maxRetry; i++ {
```

Comment:
Retrying in tight loops is problematic because:

1. it's like to fail if the issue is transient and the retry immediate
2. it floods the API server with requests

I'm honestly not sure we should retry and instead fail the recovery to let another replica take over. If we do retry, we need to queue the failed replicas and retry with a delay.

Reply:
I totally agree that we could have a better retry strategy. To do that I need more test data to support our solution. Since this PR was created last year and its main purpose is to recreate replica attachments. I'd like to merge it first so we can test it in a large scale and receive more feedback. In the meanwhile, I can add a TODO or create a GitHub issue to track task of improving the retry policy instead of blocking this PR.

However, for the worst case, the replica AzVolumeAttachment can't be recovered. When pod fails, it doesn't have replica attachment which may slow down our failover time. Then it seems like it may block us from launch which is not acceptable. If that's the case, l would agree we need a better retry strategy before merging this PR.

Let me know your thoughts.

Reply:
I would recommend dropping the retries in this PR. As I mentioned, retrying immediately is likely to fail anyway and I'd rather avoid the potential for a tight retry loop to overwhelm API server. Let's continue to discuss how to handle failures on recovery to come up with a good solution.

Reply:
How about use the same way as we do for `UpdateCRIWithRetry`?
```
err = retry.OnError(
		wait.Backoff{
			Duration: consts.CRIUpdateRetryDuration,
			Factor:   consts.CRIUpdateRetryFactor,
			Steps:    consts.CRIUpdateRetryStep,
			Cap:      consts.DefaultBackoffCap,
		},
		isRetriable,
		conditionFunc,
	)
```

Reply:
I believe that to be a workable solution.